### ISSUE TITLE
[BUG] Memoryview pure python type hints don't conform to standard type spec

### ISSUE THREAD

**Issue by mwaddoups** [cython/cython#6272]
### Describe the bug

The pure python mode of Cython is excellent, and should make it much easier to adopt cython code into Python codebases. However, there are a number of issues with the design choices around using type hints - specifically when using VSCode (and the Pyright type checker). Mypy also has similar issues.

I think this is something we should care about. For a new user using VSCode, copying any cython example with the current library throws immediate "unknown attribute" errors that will definitely put people off. But this has already been somewhat improved in #4327 - so the next release should have a better new user experience. But I propose the API is further changed.

However, the type hints and type stubs currently in use in Cython don't appear to conform to standard type specs, and the type stubs in the existing repo do not type check correctly.

Consider the example below (ran against `master`)

### Code to reproduce the behaviour:

```python
import cython
import numpy as np

@cython.ccall
def some_func_double(x: cython.double[:]) -> cython.double[:]:
    y: cython.double[:] = np.zeros(N, dtype=np.float64)
    return x
```

### Expected behaviour

With pyright, this produces the below output

```
<file>:6:39 - error: Expected no type arguments for class "float" (reportInvalidTypeArguments)
<file>:6:60 - error: Expected no type arguments for class "float" (reportInvalidTypeArguments)  
<file>:7:22 - error: Expected no type arguments for class "float" (reportInvalidTypeArguments)
<file>:7:27 - error: Expression of type "NDArray[floating[_64Bit]]" is incompatible with declared type "double"
    "ndarray[Any, dtype[floating[_64Bit]]]" is incompatible with "float" (reportAssignmentType)
<file>:7:38 - error: Cannot access attribute "shape" for class "double"
    Attribute "shape" is unknown (reportAttributeAccessIssue)
```

With mypy, this code doesn't even type check!
```
<file>:1: error: Skipping analyzing "cython": module is installed, but missing library stubs or py.typed marker  [import-untyped]
```

I can fix this by changing the `import cython` to `import Cython.Shadow as cython`. Then, `mypy` produces a similar set of errors:

```
test_cython.py:6: error: "float" expects no type arguments, but 1 given  [type-arg]
test_cython.py:6: error: Invalid type comment or annotation  [valid-type]
test_cython.py:6: note: did you mean to use ',' instead of ':' ?
test_cython.py:7: error: "float" expects no type arguments, but 1 given  [type-arg]
test_cython.py:7: error: Invalid type comment or annotation  [valid-type]
test_cython.py:7: note: did you mean to use ',' instead of ':' ?
test_cython.py:7: error: Incompatible types in assignment (expression has type "ndarray[Any, dtype[floating[_64Bit]]]", variable has type "float")  [assignment]
test_cython.py:7: error: "float" has no attribute "shape"  [attr-defined]
Found 6 errors in 1 file (checked 1 source file)
```

### OS

Linux

### Python version

3.12

### Cython version

3.0.10, and master (3.1.0a0)

### Additional context

There are a few issues here:

1. In `Shadow.pyi`, `cython.double` is set to equal `builtins.double`. This type doesn't take any type variables, producing the error.
2. The use of slices (e.g. `cython.double[:]`, `cython.double[::1]`) is not recognized as a valid type parameter specification. As far as I can see, there is no way to hack this into the type system as the implementation. So even if we took type variables, these would not be valid.
3. The use of the type variable syntax `cython.double[:]` suggests that the type is a subset of `cython.double`. This means that external code which calls these functions with (for example) a numpy array incorrectly fails a type check.

### Suggestions for solving

In the short term, I suggest that in the type stubs we type all cython types as `typing.Any`. This at least makes it clear that the onus is on the caller to check whether it's `cython.double` or `cython.double[:]` or `cython.double[::1]`, etc. 

I made and tested this change. Although `cython.double[:]` is not valid type syntax, it passes for now in Pyright (as I guess it doesn't check the details of a `typing.Any` type.

The suggestion for users would then be to write a `.pyi` file alongside your `.py` file to define the Python API.

I attempted to write a minimal set of stubs that would at least pass type checking.

In the longer term, I would suggest that cython introduces a more type-hint standard way of specifying these types. For example, `MemoryView[*T]`, `CContig[T]` etc. that still translates correctly into the appropriate cython type, but would be much easier for type checkers to check against. So the code becomes something like

```python
import cython
import numpy as np

@cython.ccall
def some_func(x: MemoryView[cython.double]) -> MemoryView[cython.double]:
    ...

@cython.ccall
def some_2d_func(x: MemoryView[cython.double, CContig[cython.double]]) -> MemoryView[cython.double, CContig[cython.double]]:
    ...
```

This should provide a much nicer API.

Alongside this, there are some noted minor bugs in `master`, including
- `@cython.nogil` does not type check
- `cython` does not have a `py.typed` marker for `mypy` to use (but `pyright` figures it out).

If you like this idea I'd be happy to look at writing a PR. I'd need a bit of help on exactly what should typecheck to memoryview.

**Comment by da-woods** [comment]
I'm not convinced by your proposed replacement memoryview syntax. It seems less readable than what we have. It also implies that you can have a memoryview like `MemoryView[cython.double, CContig[cython.int]]`

Mostly this is just confirming my pre-existing prejudices against the whole Python typing ecosystem: it's verbose, fiddly, and it infects everything.

Personally, I'd be happier with your preferred short-term solution (type everything as `Any`) as the permanent long-term solution. But I suspect I should just leave this to someone who likes and understands typing...

--------------------------------------------

> Alongside this, there are some noted minor bugs in master, [...]

Definitely do a PR for these if you have time.

**Comment by mwaddoups** [comment]
Haha - the typing system seems to be such a love it or hate it thing! For me, it's totally brought me back to Python - but mainly because of the toolchains it enables (particularly intellisense and reducing the need for testing). I think this should be the value of pure python mode - you can still use code formatting / checking tools and the full Python type checking ecosystem that your project needs, but just add a few decorators and get going.

I do think right now pure python mode is in a bit of an odd place - it looks like Python, but is interpreted slightly differently by the Cython compiler and not totally understood by Python side type checkers. And whilst the `typing.Any` hack works for now (essentially short-circuiting the type checkers), there are some functions that seem impossible to type.

For another example I came across, `SomeStruct = cython.struct(x=x, y=y)` returns a dynamic class that is also a return type. As far as I can tell, no type checker supports using a variable as a type annotation, so the suggested use as both a return type and constructor will always throw type errors. This could be solved by allowing an alternative syntax

```python
@cython.struct
class SomeStruct:
    x: cython.int
    y: cython.int
```

Similarly, type checkers have no way of knowing that `__cinit__` will get called on extension types, etc.  

You're totally right that the syntax I proposed doesn't really fit, and it is certainly verbose. I'm also not a type expert! I'll play about with it a bit and see what I can figure out. Maybe there will be some way to get the current syntax to type correctly using metaclasses or some other python magic that is unfamiliar to me.

**Comment by da-woods** [comment]
> I do think right now pure python mode is in a bit of an odd place - it looks like Python, but is interpreted slightly differently by the Cython compiler and not totally understood by Python side type checkers. And whilst the `typing.Any` hack works for now (essentially short-circuiting the type checkers), there are some functions that seem impossible to type.

Yes - it is a bit of a mix. Some of it predates annotation typing completely, things like memoryviews are adapted from our existing Cython type syntax. And some of it makes sense to run uncompiled where as some of it is sufficiently tied to C types that it'll never make sense to run without Cython.

> [...] struct [...]

The class-version of struct probably would be an improvement. Although implementing more options makes things harder to test.

> Similarly, type checkers have no way of knowing that `__cinit__` will get called on extension types, etc.

Why is this an issue?

-----

Going back a bit

```
<file>:7:27 - error: Expression of type "NDArray[floating[_64Bit]]" is incompatible with declared type "double"
    "ndarray[Any, dtype[floating[_64Bit]]]" is incompatible with "float" (reportAssignmentType)
```

This seems like a hard problem to solve without just using `Any`.
